### PR TITLE
Feature/discouraged iframe

### PR DESCRIPTION
**Pull request by ernilambar**
Fixes https://github.com/WPTRT/WordPress-Coding-Standards/issues/22

### CODE REVIEW

File: WordPress/Tests/Theme/DiscouragedIframeUnitTest.inc [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<?php
+
```
Comment (grappler) [inline WordPress/Tests/Theme/DiscouragedIframeUnitTest.inc]: We should have an example of an iframe in a PHP string.

File: WordPress/Sniffs/Theme/DiscouragedIframeSniff.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	 *                                        in the stack passed in $tokens.
+	 *
+	 * @return void
+	 */
+	public function process( PHP_CodeSniffer_File $phpcsFile, $stackPtr ) {
+		$tokens = $phpcsFile->getTokens();
+		$token  = $tokens[ $stackPtr ];
+
+		if ( preg_match( $this->iframe_regex, $token['content'] ) > 0 ) {
+			$phpcsFile->addWarning( 'iframes are sometimes used to load unwanted adverts and code on your site.' , $stackPtr, 'DiscouragedIframe' );
```
Comment (grappler) [inline WordPress/Sniffs/Theme/DiscouragedIframeSniff.php]: What about "iframes can be used to load unwanted adverts and code. Check the iframe content."?

File: WordPress/Tests/Theme/DiscouragedIframeUnitTest.inc [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+?>
+<iframe src="http://example.com" > // Warning.
+Sometime iframe is not harmful. // OK.
```
Comment (grappler) [inline WordPress/Tests/Theme/DiscouragedIframeUnitTest.inc]: What about "Sometimes iframes are not harmful."

File: WordPress/Tests/Theme/DiscouragedIframeUnitTest.inc (hunk 1 above)
Comment (grappler) [inline WordPress/Tests/Theme/DiscouragedIframeUnitTest.inc]: Sorry, I was not very clear. What I meant was `echo '<iframe src="http://example.com" >'`

**Comment by grappler** [comment]
I was thinking if there was a way to reduce the number of warnings that are generated.

Would it make sense to add a few URLs that we would whitelist like YouTube? Or discuss making this an error and only allowing certain URLs to be used.

I found a few place where only the opening tag was used for something else. It would be good not to get these false positives.
```php
$accesspress_store_widgets_allowed_tags = '<iframe>';
return strip_tags($new_field_value, $accesspress_store_widgets_allowed_tags);
```

This is from [Hybrid Core](https://github.com/justintadlock/hybrid-core/blob/master/inc/class-media-grabber.php#L443).
```php
$_html      = strip_tags( $html, '<object><embed><iframe><video>' );
```

**Comment by ernilambar** [comment]
I believe, white-listing the URLs is daunting and continuous task. May be we can allow `<iframe>` in regex?

**Comment by grappler** [comment]
This regex `\<iframe[^>]*src[^>]*>\Us` should only check for the iframes with a URL. If we have time we could discuss it in the TRT meeting today otherwise the one next week.

File: WordPress/Sniffs/Theme/DiscouragedIframeSniff.php (hunk 2 above)
Comment (jrfnl) [inline WordPress/Sniffs/Theme/DiscouragedIframeSniff.php]: `code` in that phrase, includes virusses and malware being downloaded in the background.
Are you sure you want reviewers to get virus ridden computers by checking the iframe content ? :heart_eyes:

**Comment by jrfnl** [comment]
@grappler I would consider the example code you showed invalid anyway. They should be using the more comprehensive `wp_kses()` not `strip_tags()`.

**Comment by grappler** [comment]
> I would consider the example code you showed invalid anyway. They should be using the more comprehensive wp_kses() not strip_tags().

I think that would be for another sniff. Perhaps #118 The current sniff would trip if I have a string with `<iframe>` in it.

I ran the sniff with the regex `/<iframe[^>]*src[^>]*>/Us` on all themes that have been updated in the last two years and found iframes with the following URLs:
- https://www.google.com/maps/
- https://www.google.es/maps
- https://player.vimeo.com/video/
- www.facebook.com/plugins/like.php
- www.facebook.com/plugins/follow.php
- http://fast.wistia.com/embed/iframe/
- https://www.youtube.com/embed
- https://chameleonthemes.net/docs (I don't think this URL should be allowed)

I think this would be a good list to start with a whitelist.

At the same time I found a theme doing a iframe regex `/<iframe.*?src=".*?(vimeo|youtu\.?be).*?".*?<\/iframe>/`

**Comment by ernilambar** [comment]
May be we can allow these links for embedding and restrict others. https://codex.wordpress.org/Embeds

**Comment by vovafeldman** [comment]
I think that maybe instead of whitelisting domains you'd want to take the opposite approach of blacklisting domains that violate the guidelines with malicious stuff (and maybe ban those users?). 

Also, there should be a distinguishment between WP Admin iframes (e.g. embedding something like Help Scout Docs) vs. site front iframes.

As Otto mentioned in a slack conversation, there are way more sophisticated ways to do malicious things then using iframes.

A manual review approach is requiring a PHP/JS-Doc right above the iframe inclusion to explain the use-case validity.

Example of WP Admin iframe usage is that a theme dev can add a tab in the settings page to showcase different add-ons for the theme. The benefit of using an iframe is that there's no need to release a theme update when the dev wants to add a new add-on. This page could be implemented with API calls as well. In both cases, data sent to 3rd party server without any malicious intentions (could be prompted with an opt-in mechanism). Sometimes, if the page is heavy on requests, it can be way more efficient to use an iframe to prevent a ping-pong of many API requests.

**Comment by grappler** [comment]
Thank you @vovafeldman for joining the discussion. It is appreciated!

> I think that maybe instead of whitelisting domains you'd want to take the opposite approach of blacklisting domains that violate the guidelines with malicious stuff (and maybe ban those users?).

The problem that I see with this is that end up being reactive instead of preventive. We cannot do anything for the user who have installed the update with the malicious iframe.

> Also, there should be a distinguishment between WP Admin iframes (e.g. embedding something like Help Scout Docs) vs. site front iframes.

In automation this is not so easy as it is difficult to differentiate between the two. Help Scout Docs could also be easily whitelisted though I do find a link to documentation better than having all of the documentation in the admin area.

> As Otto mentioned in a slack conversation, there are way more sophisticated ways to do malicious things then using iframes.

My concern here is that the content can be drastically changed from an external resource. Either the theme author is malicious and adds advertising in the iframe or after a few years the domain name expires and someone else buys it and uses the iframe maliciously.

 > Example of WP Admin iframe usage is that a theme dev can add a tab in the settings page to showcase different add-ons for the theme.

I think this would be a bit borderline. I am not sure how many of the key reviewers would allow this.

> This page could be implemented with API calls as well. In both cases, data sent to 3rd party server without any malicious intentions (could be prompted with an opt-in mechanism).

With the API all of the data that is returned should be escaped on output so it is a lot more difficult to drastically change the data that the API returns. There is no way to escape the content in the iframe.

**Comment by joyously** [comment]
Is this sniff supposed to allow themes to provide a page template for users to display an iframe?

**Comment by grappler** [comment]
> Is this sniff supposed to allow themes to provide a page template for users to display an iframe?

I don't see how a page template to display an iframe is needed? How would that work? The sniff is checking static code and not WordPress features.

**Comment by joyously** [comment]
A theme that I use a lot has a page template specifically for iframes. I think it does not show the sidebars as its other templates would. But now that I think about it, the theme code would not be caught by your sniff because the iframe itself is not output by the theme (it would be in the content).

**Comment by grappler** [comment]
Documenting for future reference that https://developer.wordpress.org/reference/functions/wp_oembed_get/ can be used too.

**Comment by vovafeldman** [comment]
@grappler thanks for your comments.

1. I think that the iframes "threat" is mainly in the site front, therefore, I agree with the whitelist approach for that case.

2. in terms of security I don't see a difference between an iframe and API calls in WP Admin. Modern browsers block cross domain DOM manipulation unless explicitly permitted on the iframe's containing site as a server policy. If a developer wants to do malicious things they will have to be coded on the theme side anyways to consume whatever the iframe/API is passing.

I would argue that WP Admin iframes are safer than API calls due to the cross domain limitation.

3. There are automated ways to distinguish between a WP admin iframes and site front iframes. One solution from the top of my head is adding a structured way for iframes inclusion that will simply the automated identification whether the iframe is added within WP Admin or site front. E.g. Adding a wrapper method for WP Admin inclusion that will look something like:
```
function include_admin_iframe($iframe_html){
    if (is_admin()){
        echo $iframe_html;
    }
}
```

That's just one approach. There are more sophisticated and comprehensive ways to do it with a PHP code parser (we have an extensive experience working with that at Freemius since we wrote a pre-compiler).

4. As far as I know, the top themes on .org are freemium because there are teams behind them with some business model in place. I think that as a community we should not discourage devs to upsell their paid products which eventually may cause those vendors to bail from .org (example: WPML). Obviously, it shouldn't turn to the wild wild west, so there should be a sheriff in town. But adding tasteful promotional/marketing content that doesn't do malicious things (I think) is mutually beneficial. 

5. You mentioned that a link to Help Scout Docs is better, in my opinion and based on my experience, users prefer to have it all in their WP admin dashboard, I also think it's a better UX to have the small drop at the bottom with the docs search. I'm not trying to argue here, just saying that flexibly is the one of the main powers of WordPress, I think that we should empower devs and not restrict them due to a bad behavior of very small group of malicious devs that doesn't follow the rules.

6. I'm a man of data, out of curiosity, how many malicious iframe cases the reviewers have seen in the past 12 months?

**Comment by grappler** [comment]
> I think that the iframes "threat" is mainly in the site front

I think an expired domain that shows inappropriate ads can affect the backend too.

> There are automated ways to distinguish between a WP admin iframes and site front iframes. ... There are more sophisticated and comprehensive ways to do it with a PHP code parser (we have an extensive experience working with that at Freemius since we wrote a pre-compiler).

With your example it require theme developers to write extra code to do this. The code needs to work on any code and theme and not just on a standard code. I would be interested hearing more about your a PHP code parser as we need to additional checks that PHPCS cannot handle.

> I'm a man of data, out of curiosity, how many malicious iframe cases the reviewers have seen in the past 12 months?

We don't really have this data as this review results are not in a searchable. I can tell you I only found one theme that was used an iframe for their docs and their domain name now redirects to a new one.

My goal is to remove as much manual review as possible. I would rather make the checks stricter and make an exception for every x themes then have the check too loose. Maybe we could disable certain checks for themes. By reducing the number of manual review needed we can reduce the time it takes to review a theme.

**Comment by grappler** [comment]
@ernilambar After todays team meeting we will be banning all iframes. Can you update the PR https://wordpress.slack.com/archives/themereview/p1487097663010631

**Comment by jrfnl** [comment]
Silly question maybe - but what about `<frame>` elements ? Not supported by XHTML, but supported in older HTML versions

**Comment by jrfnl** [comment]
P.S.: Sorry for being slow to review this one, traveling ATM, but will try to get to it soon.

**Comment by ernilambar** [comment]
Since we are restricting iframe altogether, may be we should rename sniff class and test.

**Comment by grappler** [comment]
> Since we are restricting iframe altogether, may be we should rename sniff class and test.

Makes sense.

What may be still interesting is to include a blank whitelist that other could override with a their custom ruleset.

File: WordPress/Tests/Theme/DiscouragedIframeUnitTest.inc (hunk 1 above)
Comment (jrfnl) [inline WordPress/Tests/Theme/DiscouragedIframeUnitTest.inc]: And what about an additional case like ? (and some variations on this)
```php
echo '<iframe src=' . esc_url( $url ) . '/>
```

File: WordPress/Sniffs/Theme/DiscouragedIframeSniff.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	 */
+	protected $iframe_regex;
+
+	/**
+	 * Returns an array of tokens this test wants to listen for.
+	 *
+	 * @return array
+	 */
+	public function register() {
+		$this->iframe_regex = '/<(iframe)[^>]*>/';
```
Comment (grappler) [inline WordPress/Sniffs/Theme/DiscouragedIframeSniff.php]: This code `strip_tags( $html, '<object><embed><iframe><video>' )` would fail if we use `strpos()`

Comment (jrfnl) [inline WordPress/Sniffs/Theme/DiscouragedIframeSniff.php]: @grappler If by "fail", you mean: would result in an error being thrown: true, but the same goes for the current regex, which is why I suspect the regex will still change ;-)

**Comment by joyously** [comment]
> Should this sniff also check for CSS rules styling <iframe> elements ? This might be an indicator that there is an <iframe> hidden somewhere in the code.

Definitely not. User content can contain iframes, and those would need styling such as max-width.

**Comment by jrfnl** [comment]
@ernilambar Hi Nilambar, are you still interested in finishing off this sniff ? If so, would you be willing to work with me over the next few weeks to get it done ?
If not, would you be ok if I updated the PR ?

Updates needed:
* PHPCS 3.x compatibility updates - see #145 
* Updates based on last review

**Comment by ernilambar** [comment]
@jrfnl Thanks for the reply. I wont be able to complete the PR as I am not quite familiar with new PHPCS. You can take over the Pull Requests and update as required.

**Comment by jrfnl** [comment]
@ernilambar Thanks for your quick response. I'll work on this later today.

**Comment by jrfnl** [comment]
I've rebased & updated the sniff. I've also squashed the original commits into one.

This should now be ready for a final review & merge.

Update notes:
* Updated for PHPCS 3.x and WPCS 0.14.x compatibility
* Renamed the sniff to "ForbiddenIframe"
* Updated based on all the feedback found in the complete thread:
    * Put the regex in a class constant instead of having it in a property and update it:
        - Remember the match so it can be used in the error message.
        - Demand at least one whitespace between the `iframe` keyword and whatever comes next and demand that there is at least one attribute.
            This prevents false positives when the tag is used on its own - which is safe anyway -.
        - Do a case-insensitive match.
    * Correct the handbook link.
    * Updated the error throwing:
        - Corrected the grammer of the error message.
        - Add "Found: ..." to the error message to make life easier on the reviewers.
        - Removed duplicate info from the error code. Any error code also contains the sniffname.
    * Update the unit tests:
        - Remove tests which don't actually test anything new.
        - Add tests for the examples given by @grappler.
        - Add test for detection within heredoc syntax as was added by the switch to using the `$textStringTokens` within the `register()` method.
        - Add test for uppercase (old-school) HTML tags.

Feedback I've not actioned:
* Detecting the `<frame>` or `<embed>` tags as it was unclear whether that was desired or not.
* Adding a URL whitelist. As `iframe`s are now [blanket forbidden](https://github.com/WPTRT/WordPress-Coding-Standards/pull/109#issuecomment-279801004), the TRTCS sniff does not need it.

**Comment by vovafeldman** [comment]
Hey guys, for some reason I was not getting updates from GitHub on this one. Anyways, I'd like to give a reminder of the chain of events, as well as suggest a less restrictive alternative, as well as to show how is it more aligned with Matt's direction of the themes review process.

## Storyline
- Back on Feb 14, 2017, there was a TRT slack meeting (which I participated in) discussing whitelisting of a selected group of domains for iframes.
- During the meeting there was an escalation which partly started as a half-joke, to ban all iframes. The main reasoning was that [it would be easier from a reviewer perspective](https://wordpress.slack.com/team/U02RWHZQR).
- The meeting notes can be found here: https://make.wordpress.org/themes/2017/02/13/agenda-for-2017-february-14/
- I disagreed (and still disagree) with the decision and from my point of view, it doesn't make any sense that a decision with such a huge impact will be decided by a group of 8-10 people. I understand that reviewing a theme is a hard job, but we also have to think about UX and the community. This decision means banning explainer YouTube videos for WordPress.org themes - that's terrible. And that's only one use-case.
- Due to my disagreement, I contacted @grappler after the meeting, told him that I think the decision doesn't really represent the community well, and asked him "what can I do?". He suggested collecting other opinions to showcase that there's a need for iframes. And that's exactly what I did.
- I started a [public poll on AWP](https://www.facebook.com/groups/advancedwp/permalink/1404605369601673/), and out of 110 votes, **only 13.6% voted to ban iframes, while 85.4% pro iframes** (74.5% voted to allow iframes from a whitelist). 110 people are not the whole community of course, but it's 10x more people than what was in the TRT meeting (including a bunch of community leaders).
- I then connected @grappler to show the results as well as added a [public comment](https://make.wordpress.org/themes/2017/02/13/agenda-for-2017-february-14/#comment-42858) on the meetings' comments.
- Finally, I discussed the topic with Ulrich during WordCamp Europe, covering our intended usage of iframes in the theme settings page, and we got an approval to use it.

## Suggestion
I suggest to get back to the original proposal and instead of banning all iframes, allow iframes with the following restrictions:
* iframes should *not* be included in the site front.
* A whitelist of domains that can be included with an iframe on the main theme's settings page (YouTube and Vimeo are good examples).
* Allow iframes on the theme setting's page, only if they are not loaded right at the main settings page, but actually require the admin to click on one of the tabs (I don't think those iframes should be restricted to a whitelist). As well as requiring those iframes/pages to include a header letting the user know it's an iframe and where is it running from, to avoid any confusion that it's part of WordPress.

## Addressing Concerns & iFrame Security
The main concern of allowing an iframe is that an expired domain can be repurchased and then show inappropriate/spammy content:
- Disallowing iframes on the site front means that even if that happens, the site visitors will not be affected by it.
- I think that we can all agree that youtube.com/maps.google.com/vimeo.com... will never show spam, or at least not in the next 10-20 years, which is more than the lifespan of most websites.
- For non-whitelisted iframes, if the iframe is only loaded in the non-default setting page (i.e. in an internal tab), there's no difference if that tab is just a link to an external page or if it's loaded inside the WP Admin as an iframe. If the domain now shows spammy content, the user will still see a new page with spammy content. There's also no difference in security. The same data that is passed to the iframe can be passed to the external page. The main difference is in the UX: when loading the content the WP Admin, we can offer a more streamlined UI and experience without dragging the admin to another website. Users love to keep everything in their WP Admin.

## Delegation to Users
Since reviewing themes is indeed hard and time-consuming, I think that some of the review process aspects can/should be delegated to theme users. Which is very much aligned with Matt's direction:
> Matt is interested in seeing how the repository reacts if we remove the manual review process and switch to a process where user feedback helps rank the themes.
Source: [Chat with Matt about the future of theme repo](https://make.wordpress.org/themes/2017/03/27/chat-with-matt-about-the-future-of-theme-repo/)

If a theme developer will include a spammy iframe (or the domain will expire and then it will become spammy), users can smell spam from miles and will cast their dissatisfaction as a 1-star review on the repo. If the iframe is spammy, there will be enough people that will give it low reviews which will affect the reputation of the theme, the theme developer, and the installs. Users do read reviews, or at least look at the avg. rate before they install something.

This will make the review process much easier and robust, and will also replace a one-man-opinion-based decision to a crowdsourced feedback. Some content may be perceived as offensive by one person due to their geolocation and culture, while the same content is considered as non-offensive by other cultures. Therefore, in my humble opinion, I don't think that the TRT should review the "content" of a theme, but restrict the review process to the code, practices, and security.

## Summary
Me, and [94 additional members of our community](https://www.facebook.com/groups/advancedwp/permalink/1404605369601673/) agree that banning all iframes for wp.org themes is not the right decision. It's too restrictive, will negatively affect creativity, as well as having a huge negative effect on UX (e.g. no explainer videos in theme settings). As described above, there is a solution/plan to keep iframes without adding more burden on the reviewers (actually simplifying the review process), while still protecting the wp.org community. I'm more than happy to discuss this further and help with the technical implementation of automating the suggestions into the theme check. My wp.org slack username is `vovafeldman`, feel free to ping me.

**Comment by dingo-d** [comment]
Great feedback @vovafeldman, if we could only get those people on facebook to be more active on the TRT channel and the review process, that would really help boost the popularity of the TRT :) (and shorten the waiting time on the themes in the queue).

I have one possible issue as to how to detect if the iframe is on admin or not. Sure you can say: ok, `is_admin()` check should be present, but what if you use an iframe inside of a template file that is included in the theme, and the `is_admin()` check is not present inside the template, but outside of it. How should the PHPCS check for that? PHPCS is just going file by file tokenizing the code and using sniffs to check the rule, it doesn't check the context of the code (as far as I'm aware).

I still think this should also be posted in the TRT channel and let admins put this in the next meeting agenda so that we can discuss it properly - the agenda should be posted on the make site and all the people on the fb are more than welcome to give their opinion in the slack channel.

Also @grappler made a good point imo

>My goal is to remove as much manual review as possible. I would rather make the checks stricter and make an exception for every x themes then have the check too loose. Maybe we could disable certain checks for themes. By reducing the number of manual review needed we can reduce the time it takes to review a theme.

**Comment by jrfnl** [comment]
@vovafeldman Just a small point of order: this is a closed PR for the automated checks part of the TRT process.
While people who were involved in this PR may get to see your comment, it will be invisible to everyone else.

This is not the right place to leave this comment or to discuss it.

To give your feedback to attention it deserves:
* If your comment regards the _implementation_ of a decision by the TRT, please open a new issue in this repo.
* If your comment regards the _decision by the TRT_, please get it added to the agenda of a TRT meeting to discuss it there.

**Comment by vovafeldman** [comment]
@jrfnl ok, thanks. There are so many channels these days, it's hard to follow if you are not in the TRT.

**Comment by vovafeldman** [comment]
@dingo-d I already suggested how to solve the identification of the iframe in the admin [here](https://github.com/WPTRT/WPThemeReview/pull/109#issuecomment-276495862). The idea is quite simple, we can introduce a new method to include iframes that will look something like that:
```
function include_admin_iframe($iframe_html){
    if (is_admin()){
        echo $iframe_html;
    }
}
```

**Comment by jrfnl** [comment]
@vovafeldman Asides from this _still_ not being the right place for this discussion - your example would not work for automated detection as there are always people who actively try to evade those checks. and those are the ones you want to "catch" with the automation, not the ones doing things right.

**Comment by vovafeldman** [comment]
@jrfnl this is true for (probably) every automation that you'll write. As long as the theme-check/sniff code is open-source, abusers will always find way to overcome it. Anyways, that's not the point right now, I will discuss it with the TRT leads.
